accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+          e.preventDefault();
+          e.stopPropagation();
+          blocklyApp.debug && console.log('in left arrow section');
+          var nextNode = node.parentNode;
+          if (node.tagName == 'BUTTON' || node.tagName == 'INPUT') {
+            nextNode = nextNode.parentNode;
+          }
+          while (nextNode && nextNode.className != 'treeview' &&
+                nextNode.tagName != 'LI') {
+            nextNode = nextNode.parentNode;
+          }
+          if (nextNode.className == 'treeview' || nextNode == null) {
+            return;
+          }
+          this.updateSelectedNode(nextNode, tree);
+          break;
+        case 38:
+          // up-facing arrow: go up a level, if possible. If not, do nothing
+          e.preventDefault();
+          e.stopPropagation();
+          console.log('node passed in: ' + node.id);
+          var prevSibling = this.getPreviousSibling(node);
+          if (prevSibling && prevSibling.tagName != 'H1') {
+            this.updateSelectedNode(prevSibling, tree);
+          } else {
+            blocklyApp.debug && console.log('no previous sibling');
+          }
+          break;
+        case 39:
+          e.preventDefault();
+          e.stopPropagation();
+          blocklyApp.debug && console.log('in right arrow section');
+          var firstChild = this.getFirstChild(node);
+          if (firstChild) {
+            this.updateSelectedNode(firstChild, tree);
+          } else {
+            blocklyApp.debug && console.log('no valid child');
+          }
+          break;
+        case 40:
+          // down-facing arrow: go down a level, if possible.
+          // If not, do nothing
+          blocklyApp.debug && console.log('preventing propogation');
+          e.preventDefault();
+          e.stopPropagation();
+          var nextSibling = this.getNextSibling(node);
+          if (nextSibling) {
+            this.updateSelectedNode(nextSibling, tree);
+          } else {
+            blocklyApp.debug && console.log('no next sibling');
+          }
+          break;
+        case 13:
+          // if I've pressed enter, I want to interact with a child
+          blocklyApp.debug && console.log('enter is pressed');
+          var activeDesc = this.getActiveDesc(treeId);
+          if (activeDesc) {
+            var children = activeDesc.children;
+            var child = children[0];
+            if (children.length == 1 && (child.tagName == 'INPUT' ||
+                child.tagName == 'BUTTON')) {
+              if (child.tagName == 'BUTTON') {
+                child.click();
+              }
+              if (child.tagName == 'INPUT') {
+                child.focus();
+              }
+            }
+          } else {
+            blocklyApp.debug && console.log('no activeDesc');
+          }
+      }
+      this.previousKey_ = e.keyCode;
+    },
+    getFirstChild: function(element) {
+      if (element == null) {
+        return element;
+      } else {
```

Comment:
Why not:

var listItems = element.getElementsByTagName('LI');
return listItems[0] || null;